### PR TITLE
Implement `am proxy` command

### PR DESCRIPTION
**Pull request by brettimus**
## Description

This PR adds support for an `am proxy` command, which would 

1. Start serving explorer, as usual (can still be configured via `--listen-address` and `--explorer-address`)
2. Optionally proxy requests to `/prometheus` to a user-specified url (configured via `--prometheus-url`)
3. **NOT** download or launch a local Prometheus

So, if we already had prometheus running on `http://localhost:8063` for some odd reason, we could run:

```
am proxy --prometheus-url http://localhost:8063
```

And `am` would not download/launch prometheus, but explorer would still be accessible and automagically connected to the Prometheus instance we're interested in connecting to.

For what it's worth, I tested that scenario, and it does indeed work.

## Where I need help

I wanted to create a handler factory, but struggled immensely with the type definition, because you can't use `impl Trait` to describe the return type for a `Fn` trait (https://github.com/rust-lang/rust/issues/99697).

What I wanted was something that looked like:

```rs
if is_proxying_prometheus {
        app = app
            .route("/prometheus/*path", any(prometheus::handler_factory(prometheus_upstream_base)))
            .route("/prometheus",  any(prometheus::handler_factory(prometheus_upstream_base)));
    }
```

Ultimately, I gave up, and opted to use inline blocks/closures to define the handlers for this functionality.

Additionally, I figured we would need to rewrite the path of any requests to `/prometheus/*` before forwarding to the external prometheus. There has to be a better/more concise way than I chose. 

Basically, this PR still needs some cleanup, api review (_is `proxy` the right command name even?_), and probably some tests.

## Checklist

- [x] Changelog updated

### CODE REVIEW

**Comment by hatchan** [comment]
> Additionally, I figured we would need to rewrite the path of any requests to /prometheus/* before forwarding to the external prometheus. There has to be a better/more concise way than I chose.

I think this is the correct way of dealing with this :+1: 

> Basically, this PR still needs some cleanup, api review (is proxy the right command name even?), and probably some tests.

I think `proxy` is the best name.

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,5 @@ target
 # This directory is created by the prometheus process
 # Could be removed once we persist the data somewhere else
 data
+
+.vscode
```
Comment (hatchan) [inline .gitignore:8]:
```suggestion
.vscode

```

File: src/bin/am/commands/proxy.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        short,
+        long,
+        env,
+        default_value = "127.0.0.1:6789",
+        alias = "explorer-address"
+    )]
+    listen_address: SocketAddr,
+
+    /// The url to which we should proxy requests to `/prometheus`
+    ///
```
Comment (hatchan) [inline src/bin/am/commands/proxy.rs:26]:
```suggestion
    /// The upstream Prometheus endpoint
```

File: src/bin/am/server.rs [hunk 3]
```diff
@@ -32,9 +38,50 @@ pub(crate) async fn start_web_server(
...
+            // 1. Get the path and query from the request, since we need to strip out `/prometheus`
+            let path_and_query = req
+                .uri()
+                .path_and_query()
+                .map(|pq| pq.as_str())
+                .unwrap_or("");
+            if let Some(stripped_path) = path_and_query.strip_prefix("/prometheus") {
+                // 2. Remove the `/prometheus` prefix.
+                let new_path_and_query =
+                    http::uri::PathAndQuery::from_maybe_shared(stripped_path.to_string())
```
Comment (hatchan) [inline src/bin/am/server.rs:66]:
```suggestion
                    http::uri::PathAndQuery::from_maybe_shared(stripped_path)
```

File: src/bin/am/commands/proxy.rs (hunk 2 above)
Comment (hatchan) [reply src/bin/am/commands/proxy.rs:26]: Or maybe actually
```suggestion
    /// The upstream Prometheus URL
```

File: src/bin/am/server.rs (hunk 3 above)
Comment (brettimus) [reply src/bin/am/server.rs:66]: so when i remove this line i get complaints:

image

Comment (brettimus) [reply src/bin/am/server.rs:66]: happening at the line 

```rs
            let path_and_query = req
                .uri()
```

Comment (brettimus) [reply src/bin/am/server.rs]: so i just removed the `.to_string()` to a different line and it resolved the error 🙃
